### PR TITLE
implement block decode in assembly

### PR DESCRIPTION
**Pull request by Zariel**
~2x faster

```
name                old time/op    new time/op     delta
UncompressPg1661-4    1.62ms ± 2%     0.61ms ± 1%    -62.14%  (p=0.000 n=10+9)
UncompressDigits-4    98.3µs ± 0%     57.2µs ± 0%    -41.76%  (p=0.000 n=10+10)
UncompressTwain-4     1.05ms ± 1%     0.39ms ± 0%    -62.61%  (p=0.000 n=10+9)
UncompressRand-4      4.16µs ± 1%     4.18µs ± 0%     +0.45%  (p=0.004 n=9+9)

name                old speed      new speed       delta
UncompressPg1661-4   367MB/s ± 2%    968MB/s ± 1%   +164.12%  (p=0.000 n=10+9)
UncompressDigits-4  1.02GB/s ± 0%   1.75GB/s ± 0%    +71.72%  (p=0.000 n=10+10)
UncompressTwain-4    371MB/s ± 1%    992MB/s ± 0%   +167.47%  (p=0.000 n=10+9)
UncompressRand-4    3.94GB/s ± 1%   3.92GB/s ± 0%     -0.45%  (p=0.006 n=9+9)

name                old alloc/op   new alloc/op    delta
UncompressPg1661-4     8.00B ± 0%      4.00B ± 0%    -50.00%  (p=0.000 n=9+8)
UncompressDigits-4     0.00B           0.00B            ~     (all equal)
UncompressTwain-4     2.40B ±100%     0.40B ±150%       ~     (p=0.077 n=10+10)
UncompressRand-4       0.00B           0.00B            ~     (all equal)

name                old allocs/op  new allocs/op   delta
UncompressPg1661-4      0.00            0.00            ~     (all equal)
UncompressDigits-4      0.00            0.00            ~     (all equal)
UncompressTwain-4       0.00            0.00            ~     (all equal)
UncompressRand-4        0.00            0.00            ~     (all equal)
```

### CODE REVIEW

**Comment by pierrec** [comment]
This is great!
I have one comment on your commit.

**Comment by Zariel** [comment]
@pierrec did you intend to leave another comment?

**Comment by pierrec** [comment]
Nope everything else looks good to me.

File: block.go [hunk 1]
```diff
@@ -30,75 +30,24 @@ func CompressBlockBound(n int) int {
...
-			expanded := dst[i:]
-			for n := offset; n <= bytesToCopy+offset; n *= 2 {
-				copy(expanded[n:], expanded[:n])
-			}
-			di += bytesToCopy
-			mLen -= bytesToCopy
-		}
-		di += copy(dst[di:di+mLen], dst[i:i+mLen])
+	di = decodeBlock(dst, src)
+	switch di {
```
[the remark this answers is not in the document]
Comment (Zariel) [reply block.go]: done

**Comment by Zariel** [comment]
implemting the shortcut decode from the C implementation brings speeds up to ~1.6GB/s on my macbook.

```
name                old time/op    new time/op    delta
UncompressPg1661-4    1.62ms ± 0%    0.37ms ± 0%   -77.31%  (p=0.000 n=9+8)
UncompressDigits-4    98.3µs ± 0%    51.8µs ± 0%   -47.37%  (p=0.000 n=10+10)
UncompressTwain-4     1.04ms ± 0%    0.24ms ± 0%   -76.87%  (p=0.000 n=10+10)
UncompressRand-4      4.19µs ± 0%    4.17µs ± 1%    -0.53%  (p=0.003 n=10+10)

name                old alloc/op   new alloc/op   delta
UncompressPg1661-4    5.60B ±100%    0.90B ±122%      ~     (p=0.053 n=10+10)
UncompressDigits-4     0.00B          0.00B           ~     (all equal)
UncompressTwain-4     2.80B ±100%    0.70B ±100%   -75.00%  (p=0.026 n=10+10)
UncompressRand-4       0.00B          0.00B           ~     (all equal)

name                old allocs/op  new allocs/op  delta
UncompressPg1661-4      0.00           0.00           ~     (all equal)
UncompressDigits-4      0.00           0.00           ~     (all equal)
UncompressTwain-4       0.00           0.00           ~     (all equal)
UncompressRand-4        0.00           0.00           ~     (all equal)

name                old speed      new speed      delta
UncompressPg1661-4   368MB/s ± 0%  1620MB/s ± 0%  +340.75%  (p=0.000 n=9+8)
UncompressDigits-4  1.02GB/s ± 0%  1.93GB/s ± 0%   +89.99%  (p=0.000 n=10+10)
UncompressTwain-4    372MB/s ± 0%  1608MB/s ± 0%  +332.37%  (p=0.000 n=10+10)
UncompressRand-4    3.91GB/s ± 0%  3.93GB/s ± 1%    +0.53%  (p=0.003 n=10+10)
```

**Comment by pierrec** [comment]
Very nice. Will try to port it to the Go function as well.

**Pull request merged**